pkg/controller/seed-controller-manager/initialmachinedeployment/controller.go

```diff
@@ -0,0 +1,243 @@
+		},
+	)
+	if err != nil {
+		r.log.Errorw("Failed to reconcile cluster", "cluster", cluster.Name, zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// If cluster is not healthy yet there is nothing to do.
+	// If it gets healthy we'll get notified by the event. No need to requeue.
+	if !cluster.Status.ExtendedHealth.AllHealthy() {
+		return nil, nil
+	}
+
+	request := cluster.Annotations[v1.InitialMachineDeploymentRequestAnnotation]
+
+	// if there is no (more) annotation, we're done
+	if request == "" {
+		return nil, nil
+	}
+
+	var nodeDeployment *v1.NodeDeployment
+	if err := json.Unmarshal([]byte(request), &nodeDeployment); err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("cannot unmarshal initial MachineDeployment request: %v", err)
+	}
+
+	nodeDeployment, err := machineresource.Validate(nodeDeployment, cluster.Spec.Version.Semver())
+	if err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("initial node deployment is not valid: %v", err)
+	}
+
+	userClusterClient, err := r.userClusterConnectionProvider.GetClient(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user cluster client: %v", err)
+	}
+
+	if err := r.createInitialMachineDeployment(nodeDeployment, cluster, userClusterClient); err != nil {
+		return nil, fmt.Errorf("failed to create initial MachineDeployment: %v", err)
+	}
+
+	if err := r.removeAnnotation(cluster); err != nil {
+		return nil, fmt.Errorf("failed to remove initial MachineDeployment annotation: %v", err)
+	}
+
+	return nil, nil
+}
+
+func (r *Reconciler) createInitialMachineDeployment(nodeDeployment *v1.NodeDeployment, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client) error {
+	datacenter, err := r.getTargetDatacenter(cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get target datacenter: %v", err)
+	}
+
+	sshKeys, err := r.getSSHKeys(cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get SSH keys: %v", err)
+	}
+
+	data := common.CredentialsData{
+		Ctx:               r.ctx,
+		KubermaticCluster: cluster,
+		Client:            r,
+	}
+
+	machineDeployment, err := machineresource.Deployment(cluster, nodeDeployment, datacenter, sshKeys, data)
+	if err != nil {
+		return fmt.Errorf("failed to assemble MachineDeployment: %v", err)
+	}
+
+	err = client.Create(r.ctx, machineDeployment)
+	if err != nil {
+		// in case we created the MD before but then failed to cleanup the Cluster resource's
+		// annotations, we can silently ignore AlreadyExists errors here and then re-try removing
+		// the annotation afterwards
+		if kerrors.IsAlreadyExists(err) {
+			return nil
+		}
```

Comment:
Will this error happen if the name of the machine deployment is not set? I think in that case we might end up with multiple machine deployments. To avoid it I have added `InitialMachineDeploymentLabel` label and `isInitialMachineDeploymentCreated` check before. It can be done in a different way if you have any idea.

Reply:
Hm I was never really sure how an initial deployment looked like. As the user can specify a name for it in the dashboard in the cluster creation wizard, I was under the impression that the initial MD will always have a name set.

Will have to investigate.

Reply:
It turns out, for MachineDeployments the `pkg/resources/machine.Deployment()` function will set a `GenerateName` on the MD if there is no user-defined name set. This will then make Kubernetes auto-generate a name by adding a unique suffix to the GenerateName and using that as the MD's name. This means every time we would apply the same annotation, we would get a new MD.

To fix this, I simply made it so that the CreateEndpoint() already sets a fixed name (using the same logic as Kubernetes itself), so that the annotation always contains a name and we can then use AlreadyExists() to perform the uniqueness check.